### PR TITLE
[Experiment] Inline noise Feb 2024

### LINKED ISSUE #331: Add support of OpusTrainer noise/inline noise modifier

### LINKED ISSUE #456: Run tests under Docker
We should create a Dockerfile that can be built locally and be close to the production environment. We should be able to either compile or pull tools like `fast_align` and `marian` to run integration tests. 

We should run tests locally and in CI under this Dockerfile.

### PR DESCRIPTION
**Pull request by eu9ene**
- produce space-tokenized alignments for teacher and student corpus to work with Tags (inline noise) modifier
- jointly train alignments for the original and back-translated corpus to improve accuracy
- separate producing shortlist to another step
- add inline noise augmentation to dataset importer using unsupervised [SimAlign](https://github.com/cisnlp/simalign) lib based on BERT to get alignments there (fast_align won't work well on a small corpus)
- adjust augmentation settings
- add tests for alignment tasks using precompiled `fast_align` and `extract_lex` from `toolchain`
- add a script to download toolchain binaries locally
- add the ability to run tests locally under Docker

See the paper: [SimAlign: High Quality Word Alignments without Parallel Training Data using Static and Contextualized Embeddings ](https://arxiv.org/abs/2004.08728)

fixes #331 
fixes #456 

[skip ci]

### CODE REVIEW

File: pipeline/alignments/generate-space-alignment.sh [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
```
Comment (gregtatum) [inline pipeline/alignments/generate-space-alignment.sh:1]: FYI: When this is ready for review, especially since this is a new bash script, I'll probably request a `run_task` test for it. This is using `fast_align` and `atools` which may or not be available via a pip install. I wonder if this would be a good place for an initial docker test, or maybe just mock the binary out again.

Comment (eu9ene) [reply pipeline/alignments/generate-space-alignment.sh:1]: yes, I wanted to introduce docker tests for this but I think it's better to write tests as a follow-up since this PR is already quite big

Comment (eu9ene) [reply pipeline/alignments/generate-space-alignment.sh:1]: Also,  nothing new is happening in the alignments, we just skip sentence piece decoding. All the critical bugs will be caught on CI. Marian fails right away when the alignments mismatch the text.

**Comment by eu9ene** [comment]
It's ready for review, but I just launched the experiment that will take a while to complete: https://firefox-ci-tc.services.mozilla.com/tasks/groups/EkGgt0YVSxu5czzeOJxBsw

File: pipeline/alignments/generate-space-alignment.sh [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+output_dir=$(dirname "${output_prefix}")
+mkdir -p "${output_dir}"
+dir="${output_dir}/tmp"
+mkdir -p "${dir}"
+
+echo "### Creating merged corpus"
+paste <(${COMPRESSION_CMD} -dc "${corpus_prefix}.${SRC}.${ARTIFACT_EXT}") <(${COMPRESSION_CMD} -dc "${corpus_prefix}.${TRG}.${ARTIFACT_EXT}") |
+  sed 's/\t/ ||| /' >"${dir}/corpus"
+
+echo "### Training alignments"
```
Comment (eu9ene) [inline pipeline/alignments/generate-space-alignment.sh:31]: Anticipating questions about this part I don't want to rewrite this in python at this point as it's mostly calling CLI commands and there will be a lot of complexity in python and it will become less readable. Also now we have a way to test the whole thing with run_task under Docker so I don't see much of a point in migration.

As for duplicating 3 lines that train alignments I also don't think extracting this will help much since it's pretty static and a standard method to train using fast_align which is unlikely to change. All other code is different across the 3 scripts.

**Comment by gregtatum** [comment]
I guess since I left a comment I'm not on the review anymore. I'll re-add myself and try to look at it tomorrow.

File: pipeline/alignments/generate-space-alignment.sh (hunk 1 above)
Comment (eu9ene) [reply pipeline/alignments/generate-space-alignment.sh:1]: I added tests that use binaries from toolchain and run locally under docker or without it if you're on linux

File: Makefile [hunk 3]
```diff
@@ -51,11 +51,33 @@ fix-all:
...
+	fi
+
+	docker build \
+		--file taskcluster/docker/base/Dockerfile \
+		--tag ftt-base .
+	docker build \
+		--build-arg DOCKER_IMAGE_PARENT=ftt-base \
+		--file taskcluster/docker/test/Dockerfile \
+		--tag ftt-test .
+	docker run -it -v $$(pwd):/builds/worker/checkouts -w /builds/worker/checkouts  ftt-test make run-tests
```
Comment (gregtatum) [inline Makefile:79]: This is failing on my macOS Ventura 13.4.1:

```
➤ make run-tests-docker
if [ $(uname -m) == 'arm64' ]; then \
	  	export DOCKER_DEFAULT_PLATFORM=linux/amd64; \
	fi
docker build \
		--file taskcluster/docker/base/Dockerfile \
		--tag ftt-base .
[+] Building 53.1s (10/10) FINISHED                                                                                                                                                                                                                                                                                                                                                                                docker:desktop-linux
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                                                                                                                                                                               0.0s
 => => transferring dockerfile: 1.75kB                                                                                                                                                                                                                                                                                                                                                                                             0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                                                                                                                                                                                                                  0.0s
 => => transferring context: 2B                                                                                                                                                                                                                                                                                                                                                                                                    0.0s
 => [internal] load metadata for docker.io/library/ubuntu:22.04                                                                                                                                                                                                                                                                                                                                                                    1.1s
 => [1/6] FROM docker.io/library/ubuntu:22.04@sha256:f9d633ff6640178c2d0525017174a688e2c1aef28f0a0130b26bd5554491f0da                                                                                                                                                                                                                                                                                                              1.8s
 => => resolve docker.io/library/ubuntu:22.04@sha256:f9d633ff6640178c2d0525017174a688e2c1aef28f0a0130b26bd5554491f0da                                                                                                                                                                                                                                                                                                              0.0s
 => => sha256:f9d633ff6640178c2d0525017174a688e2c1aef28f0a0130b26bd5554491f0da 1.13kB / 1.13kB                                                                                                                                                                                                                                                                                                                                     0.0s
 => => sha256:ca165754e2f953a4f686409b1eb5855212f42a252462c9c50bbc3077f3b9a654 424B / 424B                                                                                                                                                                                                                                                                                                                                         0.0s
 => => sha256:a50ab9f167975489853cbffd2be3bcadab3a9da27faf390ac48603c60d5c59e7 2.31kB / 2.31kB                                                                                                                                                                                                                                                                                                                                     0.0s
 => => sha256:a4a2c7a57ed8b997579870d0927433b03cfd94f5dba2153bdbcd885b5d620035 27.36MB / 27.36MB                                                                                                                                                                                                                                                                                                                                   1.0s
 => => extracting sha256:a4a2c7a57ed8b997579870d0927433b03cfd94f5dba2153bdbcd885b5d620035                                                                                                                                                                                                                                                                                                                                          0.6s
 => [2/6] RUN mkdir /builds &&     useradd -d /builds/worker -s /bin/bash -m worker &&     chown worker:worker /builds/worker &&     mkdir /builds/worker/artifacts &&     chown worker:worker /builds/worker/artifacts                                                                                                                                                                                                            0.2s
 => [3/6] WORKDIR /builds/worker/                                                                                                                                                                                                                                                                                                                                                                                                  0.0s
 => [4/6] RUN apt-get update -qq     && apt-get install -y tzdata     && apt-get install -y python3                           python3-pip                           python3-yaml                           locales                           git                           tmux                           htop                           vim                           nano                           zstd                        46.8s
 => [5/6] RUN locale-gen "en_US.UTF-8"                                                                                                                                                                                                                                                                                                                                                                                             1.0s
 => [6/6] RUN pip install zstandard                                                                                                                                                                                                                                                                                                                                                                                                1.0s
 => exporting to image                                                                                                                                                                                                                                                                                                                                                                                                             1.1s
 => => exporting layers                                                                                                                                                                                                                                                                                                                                                                                                            1.1s
 => => writing image sha256:9a06d6281301ae64f54bed0ab5aa977eef17008fb1c6b9ae39125c75e4cb5ca0                                                                                                                                                                                                                                                                                                                                       0.0s
 => => naming to docker.io/library/ftt-base                                                                                                                                                                                                                                                                                                                                                                                        0.0s

What's Next?
  View summary of image vulnerabilities and recommendations → docker scout quickview
docker build \
		--build-arg DOCKER_IMAGE_PARENT=ftt-base \
		--file taskcluster/docker/test/Dockerfile \
		--tag ftt-test .
[+] Building 12.6s (7/7) FINISHED                                                                                                                                                                                                                                                                                                                                                                                  docker:desktop-linux
 => [internal] load .dockerignore                                                                                                                                                                                                                                                                                                                                                                                                  0.0s
 => => transferring context: 2B                                                                                                                                                                                                                                                                                                                                                                                                    0.0s
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                                                                                                                                                                               0.0s
 => => transferring dockerfile: 477B                                                                                                                                                                                                                                                                                                                                                                                               0.0s
 => [internal] load metadata for docker.io/library/ftt-base:latest                                                                                                                                                                                                                                                                                                                                                                 0.0s
 => [1/4] FROM docker.io/library/ftt-base                                                                                                                                                                                                                                                                                                                                                                                          0.0s
 => [2/4] RUN apt-get update -qq     && apt-get install -y wget     && apt-get install -y pigz     && apt-get install -y unzip     && apt-get install -y zstd     && apt-get clean                                                                                                                                                                                                                                                 6.4s
 => [3/4] RUN pip install poetry                                                                                                                                                                                                                                                                                                                                                                                                   6.0s
 => ERROR [4/4] RUN chown worker:worker /builds/worker/.cache                                                                                                                                                                                                                                                                                                                                                                      0.2s
------
 > [4/4] RUN chown worker:worker /builds/worker/.cache:
0.190 chown: cannot access '/builds/worker/.cache': No such file or directory
------
Dockerfile:16
--------------------
  14 |     VOLUME /builds/worker/checkouts
  15 |     VOLUME /builds/worker/.cache
  16 | >>> RUN chown worker:worker /builds/worker/.cache
  17 |
--------------------
ERROR: failed to solve: process "/bin/sh -c chown worker:worker /builds/worker/.cache" did not complete successfully: exit code: 1
make: *** [run-tests-docker] Error 1
```

**Review by gregtatum — CHANGES_REQUESTED** [review]
I'm still reviewing but came across a Docker issue that is blocking.

* `[to review]` produce space-tokenized alignments for teacher and student corpus to work with Tags (inline noise) modifier
* `[to review]` jointly train alignments for the original and back-translated corpus to improve accuracy
* `[to review]` separate producing shortlist to another step
* `[to review]` add inline noise augmentation to dataset importer using unsupervised SimAliagn
* `[to review]` adjust augmentation settings
* `[to review]` add tests for alignment tasks using precompiled fast_align and extract_lex from toolchain
* `[to review]` add a script to download toolchain binaries locally
* `[request changes]` add the ability to run tests locally under Docker

File: utils/download-bin.sh [hunk 4]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
```
Comment (gregtatum) [inline utils/download-bin.sh:1]: Suggestion: This `utils/download-bin.sh` is a bit of a confusing name. Can we name it something like `download-toolchain.sh`

File: Makefile [hunk 5]
```diff
@@ -51,11 +51,33 @@ fix-all:
 	make black-fix
 	make lint-fix
 
+# Download binaries from Taskcluster to run tests
+# See https://firefox-ci-tc.services.mozilla.com/tasks/index/translations.cache.level-1.toolchains.v3.fast-align/latest
```
Comment (gregtatum) [inline Makefile:55]: Suggestion: `See https://firefox-ci-tc.services...` This would be better documented in the binary file.

File: utils/preflight_check.py [hunk 6]
```diff
@@ -433,30 +433,30 @@ def main(
     parameters = get_taskgraph_parameters()
     run_taskgraph(parsed_args.config, get_taskgraph_parameters())
 
-    match Choices[parsed_args.only] if parsed_args.only else None:
```
Comment (gregtatum) [inline utils/preflight_check.py:436]: Question: Why is this changed like this?

**Review by gregtatum — CHANGES_REQUESTED** [review]
I have a bit more early feedback, but plan on doing more reviewing, since the PR is larger and contains many changes.

File: taskcluster/kinds/alignments-teacher/kind.yml [hunk 7]
```diff
@@ -0,0 +1,101 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+---
+
+loader: taskgraph.loader.transform:loader
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply taskcluster/kinds/alignments-teacher/kind.yml:6]: They are required to make Tags modifier work (see docs and comments). They will be removed from the training data after the augmentation.

File: taskcluster/kinds/shortlist/kind.yml [hunk 8]
```diff
@@ -67,7 +68,7 @@ tasks:
                     export BIN=$MOZ_FETCHES_DIR &&
                     export MARIAN=$MOZ_FETCHES_DIR &&
                     export PATH=$PATH:$MOZ_FETCHES_DIR &&
-                    $VCS_PATH/pipeline/alignment/generate-alignment-and-shortlist.sh
+                    $VCS_PATH/pipeline/alignments/generate-alignment-and-shortlist.sh
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply taskcluster/kinds/shortlist/kind.yml:71]: Good question! `extract_lex` works on SentencePiece tokenized corpus and requires appropriately tokenized alignments. This is the only place where we need sp tokenized alignments so this code is a part of shortlist creation. 

Tags modifier will retokenize the space tokenized alignments into sp tokenized ones for the student model.

File: utils/preflight_check.py (hunk 6 above)
Comment (eu9ene) [reply utils/preflight_check.py:436]: I ran into issues with this under Python 3.10. The change is trivial but allows backward compatibility in case other people will run it under < 3.11.

**Comment by bhearsum** [comment]
> https://firefox-ci-tc.services.mozilla.com/tasks/Jmrh6CCOS_23BG-dNKUC9A

I see this ultimately failed because of what appears to be a bunch of spot terminations killing the `alignments-teacher` task. Do you want to switch this to run on a `-standard` instance?

**Comment by eu9ene** [comment]
> > https://firefox-ci-tc.services.mozilla.com/tasks/Jmrh6CCOS_23BG-dNKUC9A
> 
> I see this ultimately failed because of what appears to be a bunch of spot terminations killing the `alignments-teacher` task. Do you want to switch this to run on a `-standard` instance?

@bhearsum It's weird that we got 5 terminations in a row. I don't think that was the case before. Yes, we can switch but we should investigate what changed.

File: utils/download-bin.sh (hunk 4 above)
Comment (eu9ene) [reply utils/download-bin.sh]: fixed

File: pipeline/alignments/generate-space-alignment.sh [hunk 9]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+##
+# Generates space tokenized alignments to work with OpusTrainer Tags
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply pipeline/alignments/generate-space-alignment.sh]: Add a comment

File: pipeline/alignments/generate-space-alignment.sh [hunk 10]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply pipeline/alignments/generate-space-alignment.sh]: done

File: docs/opus-trainer.md [hunk 11]
```diff
@@ -23,7 +23,9 @@ Supported augmentations:
 - **Upper case** - make some sentences from the dataset upper case
 - **Title case** - use title case for some sentences from the dataset
 - **Typos** - add random typos in some words
-- **Noise** - inserts lines with random unicode noise 
+- **Noise** - inserts lines with random unicode noise
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply docs/opus-trainer.md:26]: add a link to readme

File: Makefile (hunk 3 above)
Comment (eu9ene) [reply Makefile]: It should work now

File: Makefile (hunk 5 above)
Comment (eu9ene) [reply Makefile]: moved

File: pipeline/alignments/generate-space-alignment-merged.sh [hunk 12]
```diff
@@ -0,0 +1,63 @@
+#!/bin/bash
+##
+# Generates space tokenized alignments jointly on original and backtranslated corpus to work with OpusTrainer Tags
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply pipeline/alignments/generate-space-alignment-merged.sh]: added

File: pipeline/alignments/generate-alignment-and-shortlist.sh [hunk 13]
```diff
@@ -0,0 +1,99 @@
+#!/bin/bash
+##
+# Generates alignment and lexical shortlist for a corpus.
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply pipeline/alignments/generate-alignment-and-shortlist.sh]: added more context

**Comment by eu9ene** [comment]
> > > https://firefox-ci-tc.services.mozilla.com/tasks/Jmrh6CCOS_23BG-dNKUC9A
> > 
> > 
> > I see this ultimately failed because of what appears to be a bunch of spot terminations killing the `alignments-teacher` task. Do you want to switch this to run on a `-standard` instance?
> 
> @bhearsum It's weird that we got 5 terminations in a row. I don't think that was the case before. Yes, we can switch but we should investigate what changed.

Running with a 1tb disk: https://firefox-ci-tc.services.mozilla.com/tasks/groups/DE-5HDfTSdKUsHXUUeUiOA

File: taskcluster/kinds/shortlist/kind.yml [hunk 14]
```diff
@@ -39,7 +40,7 @@ tasks:
                 - dependencies
                 - worker.env
                 - attributes
-        worker-type: b-cpu-largedisk
+        worker-type: b-cpu-xlargedisk
```
Comment (bhearsum) [inline taskcluster/kinds/shortlist/kind.yml:43]: question: What changed about this that requires more disk space? AFAICT the inputs have not changed in any way.

**Review by bhearsum — APPROVED** [review]
I reviewed only the Taskcluster parts of this, which look OK to me. I just have one question about a worker type change below.

File: taskcluster/kinds/shortlist/kind.yml (hunk 14 above)
Comment (eu9ene) [reply taskcluster/kinds/shortlist/kind.yml:43]: It's still not clear if it's the reason of failing, but I removed constant compression/decompression so now it can take more disk but less CPU and run faster which should be a good trade-off.

File: Makefile [hunk 15]
```diff
@@ -51,10 +51,33 @@ fix-all:
...
+	  	export DOCKER_DEFAULT_PLATFORM=linux/amd64; \
+	fi && \
+	docker build \
+		--file taskcluster/docker/base/Dockerfile \
+		--tag ftt-base . && \
+	docker build \
+		--build-arg DOCKER_IMAGE_PARENT=ftt-base \
+		--file taskcluster/docker/test/Dockerfile \
+		--tag ftt-test . && \
+	docker run -it --rm -v $$(pwd):/builds/worker/checkouts -w /builds/worker/checkouts ftt-test make run-tests
```
Comment (gregtatum) [inline Makefile:80]: Issue: This is still failing for me locally.

```
What's Next?
  View summary of image vulnerabilities and recommendations → docker scout quickview
poetry install --only tests --only utils --no-root
PYTHONPATH=$(pwd) poetry run pytest -vv

[Errno 2] No such file or directory: 'python'

[Errno 2] No such file or directory: 'python'
make: *** [Makefile:61: run-tests] Error 1
make: *** [run-tests-docker] Error 2
```

File: pipeline/alignments/generate-alignment-merged.sh [hunk 16]
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
```
Comment (gregtatum) [inline pipeline/alignments/generate-alignment-merged.sh:1]: I know you had a justification for not switching this to python, but it really does feel consequential enough changes to benefit from a switch. This essentially adding to technical debt at this point.

File: pipeline/alignments/generate-alignment-merged.sh [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+dir="${output_dir}/tmp"
+mkdir -p "${dir}"
+
+# train alignments on a merged corpus because fast_align is a statistical tool that benefits from a bigger corpus
+# and might not be accurate on a smaller one, for example when we have fewer back-translations
+
+# this is quite heavy on disk
+
+echo "### Decompressing original corpus to get its length"
+${COMPRESSION_CMD} -d "${original_prefix}.${SRC}.${ARTIFACT_EXT}"
```
Comment (gregtatum) [inline pipeline/alignments/generate-alignment-merged.sh:40]: This feels very inefficient, and will lead to a ton of diskspace usage. In python you can make streaming iterators that you can pass around.

File: pipeline/data/dataset_importer.py [hunk 18]
```diff
@@ -90,6 +95,29 @@ def run_cmd(cmd: List[str]):
...
 
 
+def add_alignments(corpus):
+    from simalign import SentenceAligner
+
+    # We use unsupervised aligner here because statistical tools like fast_align require a large corpus to train on
+    # This is slow without a GPU and is meant to operate only on small evaluation datasets
+
+    # Use BERT with subwords and itermax as it has a higher recall and matches more words than other methods
+    # See more details in the paper: https://arxiv.org/pdf/2311.14838.pdf
```
Comment (gregtatum) [inline pipeline/data/dataset_importer.py:105]: Did you link the correct paper here? This is all about opus trainer and the like. I don't see anything about BERT and subwords.

File: taskcluster/kinds/alignments-teacher/kind.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (gregtatum) [reply taskcluster/kinds/alignments-teacher/kind.yml:6]: > see docs and comments

I'm missing where these are. Can you link to the docs from the yml so it's easier to cross reference?

File: utils/preflight_check.py [hunk 19]
```diff
@@ -433,30 +433,30 @@ def main(
...
+    elif choice == Choices.training_config:
+        pretty_print_training_config(get_taskgraph_parameters())
+    elif choice == Choices.graph:
+        serve_taskgraph_file(
+            parsed_args.graph_url,
+            parsed_args.open_graph,
+            parsed_args.persist_graph,
+            open_in_browser,
+        )
+    elif choice is None:
```
Comment (gregtatum) [inline utils/preflight_check.py:450]: These changes look completely unrelated and should be broken out into a separate PR. I don't see any changes here.

File: utils/taskcluster_downloader.py [hunk 20]
```diff
@@ -121,7 +121,7 @@ def donwload_evals(group_id, output):
             name=artifact_name,
             queueService=queue,
         )
-        bleu, chrf, _, _ = content.tobytes().decode().split("\n")
```
Comment (gregtatum) [inline utils/taskcluster_downloader.py:124]: This change looks unrelated and should be split out into a separate PR.

File: Makefile [hunk 21]
```diff
@@ -51,10 +51,33 @@ fix-all:
...
+# Some tests work only on Linux, use Docker if running locally on other OS
 run-tests:
 	poetry install --only tests --only utils --no-root
-	PYTHONPATH=$$(pwd) poetry run pytest tests -vv
+	PYTHONPATH=$$(pwd) poetry run pytest -vv
+
+# Run unit tests locally under Docker
+# !!! IMPORTANT !!! on Apple Silicon run without poetry shell for the first time
+# as it can change `uname -m` output to x86_64 if it runs under Rosetta
+run-tests-docker:
```
Comment (gregtatum) [inline Makefile:67]: I'm running into issues getting docker working locally. This would benefit to break out into a separate PR and land first.

File: utils/download-toolchain.sh [hunk 22]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
```
Comment (gregtatum) [inline utils/download-toolchain.sh:1]: Issue: We've made the decision as a team to use python for our codebase, and this is introducing a new bash script, adding to technical debt. We should stick to the decision to use python unless we discuss as a team to change the decision.

**Review by gregtatum — CHANGES_REQUESTED** [review]
Ok, I've probably spent 5 hours total on this review so far. This PR is really too big on its own to review and safely land. It includes lots of great work, but it would be faster and safer to split into multiple PRs and land them individually.

When reviews are this big without logical commits it puts a very large burden on the reviewer. I think here we should split up the review into different parts. Anything that is not strictly alignments work should be split out into a separate PR. I left feedback on specific areas that can be split out.

I'd be happy to advise or pair on the splitting process, and work closely to give same day or next day reviews on all of the pieces so it moves quickly.

**Comment by eu9ene** [comment]
We investigated with @bhearsum and it seems the issue is OOM. I've been monitoring it through interactive task and the memory is steadily growing and reached 62% after several hours. So I guess it just reaches 100% after 9-13 hours. And we have already 256 GB worker, so it's not an option to bump it further.`fast_align` is memory intensive and I'm experimenting with switching to [eflomal](https://github.com/robertostling/eflomal/tree/master) which is supposed to more memory efficient and accurate.

I'll continue testing things here and running experiments until it works. Then I'll create a bunch of new PRs for unrelated things based on @gregtatum's feedback and either modify this one for the main logic with @gregtatum's help or create a new one. Feel free to unsubscribe until then.
